executable-spec/src/Cardano/Ledger/Spec/STS/Chain/Chain.hs

```diff
@@ -129,19 +155,62 @@ instance ( HashAlgorithm hashAlgo
       -- TODO: we will need a header transition as well, where the votes are
       -- tallied.
 
-      -- NOTE: the TRANSACTIONS transition corresponds to the BODY transition in
-      -- Byron and Shelley rules.
       let Transaction.Env
             { Transaction.updatesEnv = upE
             , Transaction.utxoEnv = utxoE
             } = transactionsEnv
-      transactionsSt' <-
+
+          (openVotingPeriods', closedVotingPeriods') = updateVotingPeriods openVotingPeriods closedVotingPeriods
+
+          updateVotingPeriods open closed =
+            let
+              -- traverse all VPs of open Map and update their status
+              updatedOpen =
+                Map.map (
+                          \vp@VotingPeriod {sipId, openingSlot, vpDuration} ->
+                            if slot > addSlot openingSlot  (vpDurationToSlotCnt vpDuration)
+                              then -- VP must close
+                                VotingPeriod
+                                   { sipId = sipId
+                                   , openingSlot = openingSlot
+                                   , closingSlot = slot
+                                   , vpDuration = vpDuration
+                                   , vpStatus = VPClosed
+                                   }
+                              else -- VP should remain open
+                                vp
+                        )
+                        open
+
+              -- extract all closed VPs from open Map
+              newClosedVPs =
+                Map.filter (\VotingPeriod {vpStatus} ->
+                                vpStatus == VPClosed
+                           )
+                           updatedOpen
+            in
+              -- insert new closed VPs into closed Map and remove closed VPs from open Map
+              ( Map.difference open newClosedVPs
+              , Map.union newClosedVPs closed
+              )
+
+      -- NOTE: the TRANSACTIONS transition corresponds to the BODY transition in
+      -- Byron and Shelley rules.
+      transactionsSt'@Transaction.St { Transaction.openVotingPeriods = ovp'
+                                        -- ovp' = returned state updated by IDEATION
+                                     , Transaction.utxoSt = _
+                                     , Transaction.updateSt = _
+                                     } <-
         trans @(TRANSACTIONS hashAlgo)
-          $ TRC ( Transaction.Env slot upE utxoE
-                , transactionsSt
+          $ TRC ( Transaction.Env slot closedVotingPeriods' upE utxoE
```

Comment:
Like we discussed yesterday, I think it is sensible for the update rules to update the opened and closed voting periods. So maybe we could add a TODO to remove this logic from the `CHAIN` STS.